add no-std support

japaric opened the pull request:
this PR adds a new API, `read_one_from_slice`, which works just like `read_one` but operates on a slice.

this PR also adds an opt-out "std" Cargo feature. all the API that operates on IO objects is behind the "std" feature. `read_one_from_slice` is available in no-std mode.

this PR is best reviewed on a commit-by-commit basis

cpu commented on src/pemfile.rs:
```diff
@@ -1,7 +1,13 @@
-use alloc::{borrow::ToOwned, format, string::String, vec::Vec};
+use alloc::{borrow::ToOwned, vec::Vec};
```
I think we'd prefer to see updated imports put in the style the Rustls repo recommends:

```
use alloc::borrow::ToOwned;
use alloc::vec::Vec;
```

cpu commented on src/pemfile.rs:
```diff
@@ -37,6 +43,56 @@ pub enum Item {
 }
 
+/// Errors that may arise when parsing the contents of a PEM file
+pub enum Error {
```
Could this change replacing the `io::Error` usage with a dedicated enum be separated out from this commit? It doesn't seem related to the `read_one_from_slice` change and isn't mentioned in the commit message.

japaric replied:
this new `Error` type is only used in the new `read_one_from_slice` API. I didn't change the signature of the existing public API as that would be a breaking change.

djc replied:
The `main` branch already has breaking changes relatively to released versions, so it's probably fine to break it more if we feel that provides the best API.

cpu commented on src/pemfile.rs:
```diff
@@ -37,6 +43,56 @@ pub enum Item {
+    Base64Decode(DecodeError),
+}
+
+impl From<DecodeError> for Error {
```
This leaks the `base64` crate into the public API no? I think one advantage of the previous approach was that it kept `base64` internal.

japaric replied:
Indeed. we can't use the usual `Box<dyn std::error::Error>` replacement because this enum has to work in no-std context. the alternatives I see are

- drop the tuple field, i.e. simply use `Error::Base64Decode`. this loses information
- use `Base64Decode(Box<dyn core::fmt::Debug>)`. that trait object is not super common but it's close to `Box<dyn Error>`. however, it cannot be converted into `Box<dyn Error>` as there's no `From` / `Into` trait that glues the two trait object types
- use `Base64Decode(String)` where the `String` is the `format!`-ed version of `Base64Decode`. unlike `Box<dyn Debug>` this can be converted into `Box<dyn Error>` using the `?` operator so that's a plus
- basically copy `base64::DecodeError` into this crate and use that. that does not expose the `base64` dependency but then you have to maintain a function that maps `base64::DecodeError` into `pemfile::DecodeError` and update it if `base64::DecodeError` (which is not marked as non_exhaustive) ever changes.

djc replied:
@ctz and I yesterday were chatting that we might want to yeet the use of the base64 crate at the first sign of trouble (since we generally haven't been that happy with it) -- maybe that's this?

cpu replied:
> use Base64Decode(String) where the String is the format!-ed version of Base64Decode. unlike Box<dyn Debug> this can be converted into Box<dyn Error> using the ? operator so that's a plus

That option seems OK to me. I suspect there's not much value in downstream code being able to programmatically determine if the data was invalid for padding, length, unknown symbol etc. In most cases I expect the error is going to be displayed to a user in string form :thinking:

japaric replied:
>  In most cases I expect the error is going to be displayed to a user in string form

I agree. this is not the case where knowing the exact error cause is going to let the user fix the input and retry the operation. most likely, malformed pem files are going to be skipped or result in a fatal error.

cpu commented on src/pemfile.rs:
```diff
@@ -36,6 +43,57 @@ pub enum Item {
+///
+/// - `Ok(None)` is returned if there is no PEM section to read from `input`
+/// - Syntax errors and decoding errors produce a `Err(...)`
+/// - Otherwise each decode section is returned with a `Ok(Some((Item::..., remainder)))` where
```
small typo?

```suggestion
/// - Otherwise each decoded section is returned with a `Ok(Some((Item::..., remainder)))` where
```

cpu commented on src/tests.rs:
```diff
@@ -1,99 +1,130 @@
-    fn check(data: &[u8]) -> Result<Vec<crate::Item>, std::io::Error> {
+    use crate::Error;
+
+    fn check_slice(mut data: &[u8]) -> Result<Vec<crate::Item>, Error> {
```
I think these helpers should be after the `#[test]`'s that rely on them, and `check_slice` after `check_both` to match [top down ordering](https://github.com/rustls/rustls/blob/main/CONTRIBUTING.md#top-down-ordering-within-modules)

cpu commented on src/tests.rs:
```diff
@@ -1,99 +1,130 @@
+        Ok(items)
+    }
+
+    fn check_io(data: &[u8]) -> Result<Vec<crate::Item>, std::io::Error> {
```
I think it might make sense to import`Item` like `Error` to avoid needing the `crate::` qualifier throughout.

cpu commented on src/pemfile.rs:
```diff
@@ -53,91 +112,127 @@ pub fn read_one(rd: &mut dyn io::BufRead) -> Result<Option<Item>, io::Error> {
-                        String::from_utf8_lossy(&line)
-                    ),
-                ));
+fn read_one_impl(
```
I think this fn should come after `read_all` since the latter is `pub`.

cpu reviewed (commented):
Thanks! Here's a first pass.

Two higher-level items:

1. It looks like this breaks `cargo test --no-default-features`.
2. Could you add CI coverage for the no-std mode?

djc commented on src/pemfile.rs:
```diff
@@ -37,6 +43,56 @@ pub enum Item {
+/// - Syntax errors and decoding errors produce a `Err(...)`
+/// - Otherwise each decode section is returned with a `Ok(Some((Item::..., remainder)))` where
+///   `remainder` is the part of the `input` that follows the returned section
+pub fn read_one_from_slice(mut input: &[u8]) -> Result<Option<(Item, &[u8])>, Error> {
```
I understand that this is trying to mirror the `read_one()` API and yet I wonder if that is actually good API -- for the conceptual action of "read one", should we treat what's represented here as `Ok(None)` as a kind of `Error` instead? And maybe we could attach this to `Item` as `Item::from_bytes()`?

japaric replied:
> should we treat what's represented here as Ok(None) as a kind of Error instead?

return something like an EOF error, say `Error::EmptySlice`? that seems reasonable at first but it may make the usage of the API a bit more clunky. right now you can write
``` rust
fn read_all_items(mut input: &[u8]) -> Result<Vec<Item>, Error> {
        let mut items = vec![];
        while let Some((item, rest)) = crate::read_one_from_slice(data)? {
            items.push(item);
            input = rest;
        }
        Ok(items)
}
```
because you can use the `?` operator on `read_from_one_from_slice`. if you were to return `Error::Eof` instead of `Ok(None)` you would have to do a more complex looking match statement to achieve the same
``` rust
fn read_all_items(input: &[u8]) -> Result<Vec<Item>, Error> {
    let mut items = vec![];
    loop {
        match crate::read_one_from_slice(data) {
            Err(Error::EmptySlice) => break Ok(items),
            Err(e) => return Err(e),
            Ok((item, rest)) => {
                items.push(item);
                input = rest;
            }
        }
    }
}
```

---

> And maybe we could attach this to Item as Item::from_bytes()?

from a readability stand point, I would expect (a) the signature of that to `fn(input: &[u8]) -> Result<Self, Error>` and (b) expect it to return an error if `input` contains more than one item or trailing bytes after one item.

I'm not opposed to having the function as an associated method on `Item` ( `fn(&[u8]) -> Result<(Self, &[u8]), Error>` does not quite fit the category of "constructor" in my mind )

djc replied:
Yeah, maybe those ideas didn't quite make sense... Hmm.

japaric commented:
I forgot to explicit answer to this:

> 1. It looks like this breaks cargo test --no-default-features.

given that the `-std` API is a strict subset of the `+std` API I don't think running `cargo t --no-default-features` in addition to `cargo t` increases coverage. given how the tests are written, I would have to `cfg` away the whole `tests` module to make `cargo t --no-default-features` work. or would you like to split all the existing tests in a `_slice` and `_io` version and run only the `_slice` ones in `-std` mode?

> 2. Could you add CI coverage for the no-std mode?

done

cpu commented:
> thanks for the review. I have addressed your comments in new commits to ease the review process. I can squash the commits when you are all happy with the final version

SGTM! Thank you.

> given that the -std API is a strict subset of the +std API I don't think running cargo t --no-default-features in addition to cargo t increases coverage. given how the tests are written, I would have to cfg away the whole tests module to make cargo t --no-default-features work.

From my perspective I think it's less about increasing coverage and more that I found it surprising that `cargo test --no-default-features` doesn't build. I expect that could confuse folks into thinking it's a bug. Gating the tests module so it doesn't fail to build seems preferable to me, even if it may mask that little testing is actually being done in that configuration. WDYT?

japaric commented:
> I found it surprising that cargo test --no-default-features doesn't build.

I've pushed a commit to make `cargo t --no-default-features` compile.

I think the expectation from the point of view of the contributor is that `cargo test` is the only thing they need to run. the reality is that aside from small library crates, project usually have complex CI that runs more than just `cargo test` so if one wants to pass CI locally one will have to read the `.yml` file and copy/paste run the commands therein. it's very few projects that take the effort of making the entire CI easy to run locally using, for example, something like an [xtask](https://github.com/matklad/cargo-xtask).

cpu reviewed (approved):
Thanks! Looks good to me once the `From` trait is fixed up.

japaric commented on src/pemfile.rs:
```diff
@@ -36,6 +44,57 @@ pub enum Item {
+}
+
+impl From<DecodeError> for Error {
+    fn from(v: DecodeError) -> Self {
```
thanks for catching that

japaric commented on src/pemfile.rs:
```diff
@@ -44,6 +103,7 @@ pub enum Item {
 /// You can use this function to build an iterator, for example:
 /// `for item in iter::from_fn(|| read_one(rd).transpose()) { ... }`
+#[cfg(feature = "std")]
 pub fn read_one(rd: &mut dyn io::BufRead) -> Result<Option<Item>, io::Error> {
```
that sounds like a change orthogonal to this PR that should be discussed in a new PR. like, do you want to provide a conversion from `pemfile::Error` to `io::Error` to ease migration from pemfile 1.0 to 2.0?

japaric replied:
there's also the catch that you can't go from `io::Error` to `pemfile::Error` and functions like `read_one` produce `io::Error`.

one could add an `Io` variant to `pemfile::Error` that sits behind the "std" Cargo feature and mark the enum as non_exhaustive, I guess, so that if an end user enables the "std" feature in their top level crate that does not break `match`-es on the enum in dependencies :thinking:

cpu replied:
> that sounds like a change orthogonal to this PR that should be discussed in a new PR.

Fair, I'll resolve this conversation then.

complexspaces commented on src/pemfile.rs:
```diff
@@ -36,6 +43,51 @@ pub enum Item {
+    let mut section = None::<(Vec<_>, Vec<_>)>;
+
+    loop {
+        let next_line = if let Some(index) = input.iter().position(|byte| *byte == b'\n') {
```
Out of curiosity, were these changes benchmarked? I was also making some changes to the implementation here and observed that without std's usage of `memchr` or the `memchr` crate, performance tanks by about ~2x in my branch due to the performance looking for line endings.

complexspaces replied:
Is that with the `std` feature disabled? It seems odd that they are so close but its not a big deal.

japaric commented:
I've squashed the commits

djc commented:
I think this the `no-std` commit has too much going on right now. IMO it would be good to have separate commits that:

1. Split `read_one_impl()` out from `read_one()`
2. Introduce the new `Error` type
3. Rename the `base64` module (why is this necessary?)
4. Introduce `read_one_from_slice()`

Additionally, I find the current order of items pretty confusing. Currently the order is kind of bottom-up (contra our conventions in other crates), but this PR then adds `read_one_impl()` at the end instead of at the top which would be more consistent.

japaric commented:
Split the commits. 

> Rename the base64 module (why is this necessary?)

in a previous version it was masking the `base64` dependency and making a `use base64::Error` import fail. does not seem necessary after removing `base64::Error` from the `Error::Base64Decode` variant

japaric commented:
I dropped the bit that disabled base64's "std" feature while splitting the commit (and CI still passed). I have updated CI (in the latest commit) to catch usage of libstd in dependencies.

djc commented on src/pemfile.rs:
```diff
@@ -57,83 +58,107 @@ pub fn read_one(rd: &mut dyn io::BufRead) -> Result<Option<Item>, io::Error> {
+        if let ControlFlow::Break(opt) = read_one_impl(next_line, &mut section, &mut b64buf)? {
+            return Ok(opt);
+        } else {
+            continue;
```
Style: I think both of these would read better as `match` expressions.

djc commented on src/pemfile.rs:
```diff
@@ -64,10 +83,30 @@ pub fn read_one(rd: &mut dyn io::BufRead) -> Result<Option<Item>, io::Error> {
+            Ok(ControlFlow::Break(opt)) => return Ok(opt),
+            Ok(ControlFlow::Continue(())) => continue,
+            Err(e) => {
+                let e = match e {
```
Style nit: if you forego the `e` binding in favor of yielding `Err(match e { ..})` directly this will format a bit more nicely.

djc reviewed (approved):
Thanks for splitting, this is much better!

japaric commented:
@djc applied your suggestions
